scanner.go

```diff
@@ -0,0 +1,285 @@
+// If a structField is nil, it means it couldn't be matched and we insert
+// a ignoreFieldType pointer instead. This means you will always get back
+// len(structFields) pointers initialized
+func generatePtrs(structFields []*r.Field) []interface{} {
+	ptrs := make([]interface{}, len(structFields))
+	for i, sf := range structFields {
+		if sf == nil {
+			ptrs[i] = &ignoreFieldType{}
+			continue
+		}
+
+		val := reflect.New(sf.Type())
+		ptrs[i] = val.Interface()
+	}
+	return ptrs
+}
+
+// setPtrs takes a list of fields and the associated pointers and sets them
+// in order to the targetStruct
+func setPtrs(structFields []*r.Field, ptrs []interface{}, targetStruct reflect.Value) {
+	for index, field := range structFields {
+		if field == nil {
+			continue
+		}
+
+		// Handle the case where the embedded struct hasn't been allocated yet
+		// if it's a pointer. Because these are anonymous, if they are nil we
+		// can't access them! We could be smarter here by allocating embedded
+		// pointers (if they aren't allocated already) and traversing the
+		// struct allocating all the way down as necessary
+		if len(field.Index()) > 1 {
+			elem := targetStruct.FieldByIndex([]int{field.Index()[0]})
+			if elem.Kind() == reflect.Ptr && elem.IsNil() {
+				continue
+			}
+		}
+
+		elem := targetStruct.FieldByIndex(field.Index())
+		if elem.CanSet() {
+			data := reflect.ValueOf(ptrs[index]).Elem()
+
+			// To preserve old behaviour, if we got a nil element back, we need
+			// to initialize it ourselves
+			switch data.Kind() {
+			case reflect.Map:
+				if data.IsNil() {
+					data = reflect.MakeMap(elem.Type())
+				}
+			case reflect.Slice:
+				if data.IsNil() {
+					data = reflect.MakeSlice(elem.Type(), 0, 0)
+				}
+			}
+
+			elem.Set(data)
+		}
+	}
+}
+
+// allocateNilReference checks to see if the in is not nil itself but points to
+// an object which itself is nil. Note that it only checks one depth down.
+// Returns true if any allocation has happened, false if no allocation was needed
+func allocateNilReference(in interface{}) error {
+	val := reflect.ValueOf(in)
+	if val.Kind() != reflect.Ptr {
+		return nil
+	}
+
+	if val.IsNil() {
+		return fmt.Errorf("pointer passed in was nil itself (not addressable)")
+	}
+
+	// Don't re-allocate if we don't need to. If the underlying element is not
+	// nil then we can avoid an alloc (only checks depth = 1)
+	switch val.Elem().Kind() {
+	case reflect.Map, reflect.Slice:
+		if !val.Elem().IsNil() {
+			return nil
+		}
+	}
+
+	// Here we unravel the underlying base type, it's just
+	// pointer turtles all the way down
+	topLevelType := reflect.TypeOf(in)
+	baseType := reflect.TypeOf(in)
+	for baseType.Kind() == reflect.Ptr {
+		baseType = baseType.Elem()
+	}
+
+	var basePtr reflect.Value
+	switch baseType.Kind() {
+	case reflect.Array, reflect.Chan, reflect.Func, reflect.Interface, reflect.Ptr, reflect.UnsafePointer:
+		return fmt.Errorf("type of kind %v is not supported", baseType.Kind())
```

Comment:
Minor suggestion, might be worth explaining why in a comment (these aren't settable).